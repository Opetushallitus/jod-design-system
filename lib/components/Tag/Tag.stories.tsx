import type { Meta, StoryObj } from '@storybook/react';

import { fn } from '@storybook/test';
import { Tag } from './Tag';

const meta = {
  title: 'Tag',
  component: Tag,
  tags: ['autodocs'],
} satisfies Meta<typeof Tag>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Selectable: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542-6833',
    },
  },
  args: {
    label: 'selectable',
    onClick: fn(),
  },
};

export const Added: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542-6866',
    },
  },
  args: {
    label: 'added',
    onClick: fn(),
    variant: 'added',
  },
};
