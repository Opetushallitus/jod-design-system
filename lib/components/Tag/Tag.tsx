type TagColors = 'secondary-2' | 'secondary-3' | 'secondary-4' | 'secondary-5';
export interface TagProps {
  label: string;
  onClick: () => void;
  variant?: 'selectable' | 'added';
  color?: TagColors;
}

export const Tag = ({ label, onClick, variant = 'selectable', color = 'secondary-4' }: TagProps) => {
  const colorMap: Record<TagColors, string> = {
    'secondary-2': 'bg-secondary-2-50',
    'secondary-3': 'bg-secondary-3-50',
    'secondary-4': 'bg-secondary-4-50',
    'secondary-5': 'bg-secondary-5',
  };
  return (
    <button
      type="button"
      className={`group flex select-none items-center rounded-xl text-tag px-4 py-2 ${colorMap[color]}`}
      onClick={onClick}
    >
      <span className="hyphens-auto text-black group-hover:underline">{label}</span>
      <span className="material-symbols-outlined pl-3 select-none text-button-md text-black" aria-hidden>
        {variant === 'selectable' ? 'add' : 'close'}
      </span>
    </button>
  );
};
