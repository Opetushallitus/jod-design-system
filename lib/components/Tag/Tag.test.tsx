import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Tag } from './Tag';

describe('Tag', () => {
  it('matches the snapshot', () => {
    const { container } = render(<Tag label="Label here" onClick={vi.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it('renders the label correctly', () => {
    const label = 'Test Label';
    const { getByText } = render(<Tag label={label} onClick={vi.fn()} dataTestId="tag1" />);
    expect(document.querySelector('[data-testid="tag1"]')).toBeInTheDocument();
    expect(getByText(label)).toBeInTheDocument();
  });

  it('calls the onClick function when clicked', () => {
    const onClick = vi.fn();
    const { getByRole } = render(<Tag label="onClick testing" onClick={onClick} dataTestId="tag2" />);
    fireEvent.click(getByRole('button'));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should not be a button and no icons are present when using presentation variant', () => {
    const { container, queryByRole } = render(<Tag label="presentation" variant="presentation" dataTestId="tag3" />);
    expect(document.querySelector('[data-testid="tag3"]')).toBeInTheDocument();
    expect(queryByRole('button')).toBeNull();
    expect(container.querySelector('svg')).toBeNull();
  });

  it('should have cursor-pointer class when variant is not presentation', () => {
    const { container } = render(<Tag label="selectable" variant="selectable" onClick={vi.fn()} />);
    const tagElement = container.firstChild;
    expect(tagElement).toHaveClass('ds:cursor-pointer');
  });

  it('should not have cursor-pointer class when variant is presentation', () => {
    const { container } = render(<Tag label="presentation" variant="presentation" />);
    const tagElement = container.firstChild;
    expect(tagElement).not.toHaveClass('ds:cursor-pointer');
  });

  it('should have cursor-pointer class by default (when variant not specified)', () => {
    const { container } = render(<Tag label="default" onClick={vi.fn()} />);
    const tagElement = container.firstChild;
    expect(tagElement).toHaveClass('ds:cursor-pointer');
  });
});
