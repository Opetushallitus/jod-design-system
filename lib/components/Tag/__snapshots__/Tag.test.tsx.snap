// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > matches the snapshot 1`] = `
<div>
  <button
    class="ds-group ds-flex ds-select-none ds-items-center ds-rounded-xl ds-text-tag ds-font-arial ds-px-4 ds-py-2 ds-bg-secondary-4-50"
    type="button"
  >
    <span
      class="ds-hyphens-auto ds-text-black group-hover:ds-underline"
    >
      Label here
    </span>
    <span
      aria-hidden="true"
      class="material-symbols-outlined ds-pl-3 ds-select-none ds-text-button-md ds-text-black"
    >
      add
    </span>
  </button>
</div>
`;
